src/stratigraphy/util/geometric_line_utilities.py

```diff
@@ -381,3 +225,101 @@ def _are_parallel(line1: Line, line2: Line, angle_threshold: float) -> bool:
         bool: True if the lines are parallel, False otherwise.
     """
     return np.abs(atan(line1.slope) - atan(line2.slope)) < angle_threshold * pi / 180
+
+
+def merge_parallel_lines_quadtree(lines: list[Line], tol: int, angle_threshold: float) -> list[Line]:
+    """Merge parallel lines that are close to each other.
+
+    Uses a quadtree to quickly find lines that are close to each other. The algorithm is more efficient than the
+    naive approach.
+
+    Args:
+        lines (list[Line]): The lines to merge.
+        tol (int, optional): Tolerance to check if lines are close.
+        angle_threshold (float, optional): Acceptable difference between the slopes of two lines.
+
+    Returns:
+        list[Line]: The merged lines.
+    """
+    indexed_lines = IndexedLines(lines)
+
+    # Create a quadtree
+    width = max([line.end.x for line in indexed_lines.hashmap.values()])
+    max_end_y = max([line.end.y for line in indexed_lines.hashmap.values()])
+    max_start_y = max([line.start.y for line in indexed_lines.hashmap.values()])
+    height = max(max_end_y, max_start_y)
+    qtree = quads.QuadTree(
+        (width / 2, height / 2), width + 1000, height + 1000
+    )  # Add some margin to the width and height to allow for slightly negative values
+    for line_index, line in indexed_lines.hashmap.items():
+        qtree.insert((line.start.x, line.start.y), data=line_index)
+        qtree.insert((line.end.x, line.end.y), data=line_index)
+
+    keys_queue = queue.Queue()
+    for key in indexed_lines.hashmap:
+        keys_queue.put(key)
+    merged_any = False
+    while not keys_queue.empty():
+        line_index = keys_queue.get()
+        if line_index in indexed_lines.hashmap:
+            line = indexed_lines.hashmap[line_index]
+            # Get all points in the quadtree that are close to the line
+            min_x = min(line.start.x, line.end.x)
+            max_x = max(line.start.x, line.end.x)
+            min_y = min(line.start.y, line.end.y)
+            max_y = max(line.start.y, line.end.y)
+            bb = quads.BoundingBox(min_x - tol, min_y - tol, max_x + tol, max_y + tol)
+
+            points = qtree.within_bb(bb)
+            merge_candidates_idx = get_merge_candidates_from_points(points, indexed_lines)
```

Comment:
I'm still struggling to understand why the recursive calls of this method are necessary. I indeed see that we are still merging more lines after the first call of this method. However, I do not understand why.

Indeed, if I replace these lines with the follow (less efficient, but intuitively equivalent) code, then all the merges already happen in the first iteration.
```
            merge_candidates_idx = []
            for idx, merge_line in indexed_lines.hashmap.items():
                if (
                    min_x - tol < merge_line.end.x <= max_x + tol and min_y - tol < merge_line.end.y <= max_y + tol
                ) or (
                    min_x - tol < merge_line.start.x <= max_x + tol and min_y - tol < merge_line.start.y <= max_y + tol
                ):
                    merge_candidates_idx.append(idx)
```
In other words, there seem to be valid merge candidates, that are not immediately found by the `qtree.within_bb` method, at least not until the next iteration, after we've reconstructed the quad tree.

Is this an issue with the quads library, or with our usage of quad trees? I don't fully understand it yet. Still needs more debugging...

Reply:
So indeed, the current quads implementation runs into problems when you try to use it with points that have the same coordinates but potentially different data. The problems start as soon as there are more than 4 points with the same coordinates, as this is the default capacity value (https://github.com/toastdriven/quads/blob/master/quads.py#L279). You can increase the capacity value when constructing the quad tree, but you never know for sure how large the value has to be, in order not to run into errors, and also this could make the quad tree less efficient.

However, in general, it should be no problem to write a quad tree implementation that supports our use case. Either you check the capacity based on unique coordinates, rather than based on total number of points. Or you make the data associated with each point a set (containing all the indexes of lines associated with that point), to which elements can be added or removed dynamically. The latter approach might actually be possible with the current quads library, I'll give it a try.

Reply:
The idea from the previous comment seems to work :). See https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/43
The implementation still needs to be documented + cleaned up a little, but I think I prefer it over the current workaround/recursion.